### PR TITLE
fix(@angular-devkit/build-optimizer): wrap es2015 class expressions for better tree-shaking

### LINKED ISSUE #14577: ClassExpression in ES2015 builds produces larger bundles
# 🐞 Bug report

### Command (mark with an `x`)

```
- [ ] new
- [x] build
- [ ] serve
- [ ] test
- [ ] e2e
- [ ] generate
- [ ] add
- [ ] update
- [ ] lint
- [ ] xi18n
- [ ] run
- [ ] config
- [ ] help
- [ ] version
- [ ] doc
```

### Is this a regression?

No, but now it effects more users as differential loading is on by default.

### Description

TypeScript ClassExpression are not being treeshaken away.

Example codes:

```ts
let AggregateColumnDirective = class AggregateColumnDirective {
    constructor(viewContainerRef) {
    }
};
AggregateColumnDirective = __decorate([
    Directive({
        selector: 'ejs-grid>e-aggregates>e-aggregate>e-columns>e-column',
        inputs: input$1,
        outputs: outputs$1,
        queries: {}
    }),
    __metadata("design:paramtypes", [ViewContainerRef])
], AggregateColumnDirective);
```

```ts
let ChipList = class ChipList extends Component {
  constructor(options, element) {
        super(options, element);
    }
};
__decorate$4([
    Property([])
], ChipList.prototype, "chips", void 0);
ChipList = __decorate$4([
    NotifyPropertyChanges
], ChipList);
```

## 🌍 Your Environment
<pre><code>
Angular CLI: 8.0.0
Node: 11.0.0
OS: win32 x64
Angular: 8.0.0
... animations, cli, common, compiler, compiler-cli, core, forms
... language-service, platform-browser, platform-browser-dynamic
... router

Package                           Version
-----------------------------------------------------------
@angular-devkit/architect         0.800.0
@angular-devkit/build-angular     0.800.0
@angular-devkit/build-optimizer   0.800.0
@angular-devkit/build-webpack     0.800.0
@angular-devkit/core              8.0.0
@angular-devkit/schematics        8.0.0
@ngtools/webpack                  8.0.0
@schematics/angular               8.0.0
@schematics/update                0.800.0
rxjs                              6.4.0
typescript                        3.4.5
webpack                           4.30.0
</code></pre>

**Anything else relevant?**

https://github.com/angular/angular-cli/pull/14263

### LINKED ISSUE #14610: ES2015 Class with decorators are not being treeshaken away
# 🐞 Bug report

### Command (mark with an `x`)

```
- [ ] new
- [x] build
- [ ] serve
- [ ] test
- [ ] e2e
- [ ] generate
- [ ] add
- [ ] update
- [ ] lint
- [ ] xi18n
- [ ] run
- [ ] config
- [ ] help
- [ ] version
- [ ] doc
```

### Description
When not using tsickle, TypeScript emits the following code for classes which have property or method decorators

```ts
export class foo {
    bar() {
    }
}
__decorate([
    decorator()
], foo.prototype, "bar", null);
```

The problematic as terser is unable to drop the above code if the Class is unused.

## 🌍 Your Environment
<pre><code>
Angular CLI: 8.0.0
Node: 11.0.0
OS: win32 x64
Angular: 8.0.0
... animations, cli, common, compiler, compiler-cli, core, forms
... language-service, platform-browser, platform-browser-dynamic
... router

Package                           Version
-----------------------------------------------------------
@angular-devkit/architect         0.800.0
@angular-devkit/build-angular     0.800.0
@angular-devkit/build-optimizer   0.800.0
@angular-devkit/build-webpack     0.800.0
@angular-devkit/core              8.0.0
@angular-devkit/schematics        8.0.0
@ngtools/webpack                  8.0.0
@schematics/angular               8.0.0
@schematics/update                0.800.0
rxjs                              6.4.0
typescript                        3.4.5
webpack                           4.30.0
</code></pre>

**Anything else relevant?**

Related to #14577

### PR DESCRIPTION
**Pull request by alan-agius4**
fix(@angular-devkit/build-optimizer): wrap es2015 class expressions for better tree-shaking 

ClassExpressions such as the below are not treeshakable unless we wrap them in an IIFE

```js
let AggregateColumnDirective = class AggregateColumnDirective {
	constructor(viewContainerRef) { }
};
AggregateColumnDirective = __decorate([
	Directive({}),
	__metadata("design:paramtypes", [ViewContainerRef])
], AggregateColumnDirective);
```

With this change we wrap the above in an IIFE and mark it as a PURE function.
```js
const AggregateColumnDirective = /*@__PURE__*/ (() => {
	let AggregateColumnDirective = class AggregateColumnDirective {
		constructor(viewContainerRef) { }
	};
	AggregateColumnDirective = __decorate([
		Directive({}),
		__metadata("design:paramtypes", [ViewContainerRef])
	], AggregateColumnDirective);

	return AggregateColumnDirective;
})();
```

With this pattern if the class is unused it will be dropped.

Note: In future we should rename `wrap-enums` to something more generic, and combine class-fold with this transformer especially considering the future fix that needs to be done for https://github.com/angular/angular-cli/issues/14610

Fixes #14577

fix(@angular-devkit/build-optimizer) wrap ClassDeclarations in an IIFE for better treeshaking

With this change we wrap ClassDeclarations inside an IIFE, also we move some code from the class fold into the wrap-enums.

This changes the below code:
```js
export class Foo {
	method() {
	}
}
Foo.bar = 'barValue';
__decorate([
	methodDecorator
], Foo.prototype, "method", null);
```

to 

```js
export const Foo = /*@__PURE__*/ (() => {
  class Foo {
    method() {
    }
  }
  Foo.bar = 'barValue';
  __decorate([
    methodDecorator
  ], Foo.prototype, "method", null);

  return Foo;
})();
```

Fixes #14610

test: add test to verify that the new es2015 class wrapping logic handles wrapping of tslib and tsickle classes

Related to https://github.com/ngrx/platform/issues/1905 and https://github.com/ng-packagr/ng-packagr/issues/1307

Fixes: #14613

### CODE REVIEW

File: packages/angular_devkit/build_optimizer/src/transforms/wrap-enums.ts [hunk 1]
```diff
@@ -411,6 +433,56 @@ function updateHostNode(
...
+ Foo = __decorate([]);
+ ```
+ */
+function findClassExpressionStatements(
+  name: string,
+  statements: ts.NodeArray<ts.Statement>,
+  statementIndex: number,
+): ts.Statement[] {
+  const classStatements: ts.Statement[] = [];
+  for (let index = statementIndex + 1; index < statements.length; ++index) {
```
Comment (alan-agius4) [inline packages/angular_devkit/build_optimizer/src/transforms/wrap-enums.ts]: Self note, change to `while`

**Comment by bgotink** [comment]
It appears classes that self-reference using `forwardRef` lead to slightly different output:

```js
var TestUnshakeableDirective_1;
export const SOME_TOKEN = new InjectionToken('some token');
let TestUnshakeableDirective = TestUnshakeableDirective_1 = class TestUnshakeableDirective {
    constructor() { }
};
TestUnshakeableDirective = TestUnshakeableDirective_1 = __decorate([
    Directive({
        selector: '[libUnshakeable]',
        providers: [
            {
                provide: SOME_TOKEN,
                useExisting: forwardRef(() => TestUnshakeableDirective_1)
            },
        ]
    })
], TestUnshakeableDirective);
export { TestUnshakeableDirective };
```

and

```js
var TestUnshakeable2Directive_1;
let TestUnshakeable2Directive = TestUnshakeable2Directive_1 = class TestUnshakeable2Directive {
    constructor(parent) { }
};
TestUnshakeable2Directive = TestUnshakeable2Directive_1 = __decorate([
    Directive({
        selector: '[libUnshakeable2]',
    }),
    __param(0, SkipSelf()), __param(0, Inject(forwardRef(() => TestUnshakeable2Directive_1))),
    __metadata("design:paramtypes", [TestUnshakeable2Directive])
], TestUnshakeable2Directive);
export { TestUnshakeable2Directive };
```

The output doesn't change for `forwardRef`s outside of the class though
<details>
<summary>code</summary>

```js
const SOME_TOKEN = new InjectionToken('some token');
const PROVIDE_SOME_TOKEN = {
    provide: SOME_TOKEN,
    useExisting: forwardRef(() => TestUnshakeableDirective)
};
let TestUnshakeableDirective = class TestUnshakeableDirective {
    constructor() { }
};
TestUnshakeableDirective = __decorate([
    Directive({
        selector: '[libUnshakeable]',
        providers: [
            PROVIDE_SOME_TOKEN,
        ]
    })
], TestUnshakeableDirective);
```

and

```js
let DoorDirective = class DoorDirective {
    constructor(lock) {
        this.lock = lock;
    }
};
DoorDirective = __decorate([
    Directive({ selector: '[libDoor]' }),
    __param(0, Inject(forwardRef(() => LockDirective))),
    __metadata("design:paramtypes", [LockDirective])
], DoorDirective);
let LockDirective = class LockDirective {
};
LockDirective = __decorate([
    Directive({ selector: '[libLock]' })
], LockDirective);
```

</details>

**Comment by alan-agius4** [comment]
@bgotink, thanks for the above, it will surly help to improve the logic to handle the above case.

File: packages/angular_devkit/build_optimizer/src/transforms/class-fold.ts [hunk 2]
```diff
@@ -6,7 +6,6 @@
  * found in the LICENSE file at https://angular.io/license
```
Comment (alan-agius4) [inline packages/angular_devkit/build_optimizer/src/transforms/class-fold.ts]: @clydin and @filipesilva should we remove the class-fold ES5 logic completely. Ie depreciate it and not use it?

I tried this and it seems that static properties are always folding an in an IIFE by typescript at least for version 2.0+

Comment (filipesilva) [reply packages/angular_devkit/build_optimizer/src/transforms/class-fold.ts]: I think we should deprecate, but not remove yet. It's still part of the public API.

**Comment by alan-agius4** [comment]
@filipesilva comments fixed. PTAL

**Comment by alan-agius4** [comment]
Should we do the class wrapping only when it's sideEffects free?

**Comment by clydin** [comment]
It should always happen to be consistent with TS ES5 class behavior and hopefully standard TS behavior in general moving forward.

**Comment by filipesilva** [comment]
I've removed myself as reviewer because I think @clydin understand this particular transformer much better than me.

**Pull request merged**